File: scripts/smoke_test_authority.py
#!/usr/bin/env python3
"""End-to-end smoke test for NuggetsAuthorityMiddleware against a live backend.

Hits the real Nuggets authority endpoint with one ALLOW-expected tool call
through the middleware. Exits 0 on success, non-zero on failure.

Use against a deployed environment (dev / staging / prod) to confirm the
SDK's request shape, auth flow, and response handling work against the
real Next.js route at /api/authority/evaluate.

Setup:
    Register the agent and download its private key from the accounts portal.
    Pre-create a delegation with NUGGETS_TOOL in its allowed_capabilities;
    note the delegation_id.

Usage:
    export NUGGETS_AUTHORITY_URL="https://accounts-dev.internal-nuggets.life"
    export NUGGETS_OIDC_ISSUER_URL="https://auth-dev.internal-nuggets.life"
    export NUGGETS_AGENT_ID="did:web:auth-dev.internal-nuggets.life:..."
    export NUGGETS_CONTROLLER_ID="did:nuggets:oidc:..."
    export NUGGETS_DELEGATION_ID="42"
    export NUGGETS_AGENT_PRIVATE_KEY="/path/to/agent-jwks.json"  # or PEM
    export NUGGETS_TOOL="your_tool_name"    # any capability in the delegation's allowed_capabilities
    export NUGGETS_TARGET="your_target"     # optional; must match the
                                            # delegation's allowed_targets when set

    python scripts/smoke_test_authority.py
"""
from __future__ import annotations

import json
import os
import sys
from types import SimpleNamespace
from typing import Any, Dict, Union

from langchain_core.messages import ToolMessage

from langchain_nuggets.middleware import MiddlewareConfig, NuggetsAuthorityMiddleware

REQUIRED_ENV = [
    "NUGGETS_AUTHORITY_URL",
    "NUGGETS_OIDC_ISSUER_URL",
    "NUGGETS_AGENT_ID",
    "NUGGETS_CONTROLLER_ID",
    "NUGGETS_DELEGATION_ID",
    "NUGGETS_AGENT_PRIVATE_KEY",
]


def fail(message: str) -> None:
    print(f"FAIL: {message}", file=sys.stderr)
    sys.exit(1)


def load_key_from_env() -> Union[str, Dict[str, Any]]:
    """Resolve NUGGETS_AGENT_PRIVATE_KEY to a value MiddlewareConfig accepts.

    A file path (PEM, JWK JSON, JWKS JSON) is handed off as-is to the
    middleware loader, which handles all three forms. Inline JSON is
    parsed and passed as a dict so MiddlewareConfig validation runs
    against a structured value rather than a raw string.
    """
    raw = os.environ["NUGGETS_AGENT_PRIVATE_KEY"]
    if raw.lstrip().startswith("{"):
        return json.loads(raw)
    return raw


def main() -> None:
    missing = [name for name in REQUIRED_ENV if not os.environ.get(name)]
    if missing:
        fail(f"missing env vars: {', '.join(missing)}")

    tool_name = os.environ.get("NUGGETS_TOOL")
    if not tool_name:
        fail("NUGGETS_TOOL must be set to a capability listed in the delegation")

    config = MiddlewareConfig(
        api_url=os.environ["NUGGETS_AUTHORITY_URL"],
        oidc_issuer_url=os.environ["NUGGETS_OIDC_ISSUER_URL"],
        agent_id=os.environ["NUGGETS_AGENT_ID"],
        controller_id=os.environ["NUGGETS_CONTROLLER_ID"],
        delegation_id=os.environ["NUGGETS_DELEGATION_ID"],
        agent_private_key=load_key_from_env(),
    )

    middleware = NuggetsAuthorityMiddleware(config)

    target = os.environ.get("NUGGETS_TARGET")
    args: Dict[str, Any] = {"userId": "smoke-test-user"}
    if target:
        args["target"] = target

    request = SimpleNamespace(
        tool_call={
            "name": tool_name,
            "args": args,
            "id": "smoke-call-001",
        }
    )

    def handler(_: object) -> ToolMessage:
        return ToolMessage(
            content=json.dumps({"status": "ok", "smoke": True}),
            tool_call_id="smoke-call-001",
        )

    print(f"Calling authority at {config.api_url}{config.authority_endpoint}")
    print(f"  agent_id      = {config.agent_id}")
    print(f"  controller_id = {config.controller_id}")
    print(f"  delegation_id = {config.delegation_id}")
    print(f"  tool          = {tool_name}")

    result = middleware.wrap_tool_call(request, handler)

    if not isinstance(result, ToolMessage):
        fail(f"unexpected result type: {type(result).__name__}")

    try:
        payload = json.loads(result.content)
    except json.JSONDecodeError:
        fail(f"result content not JSON: {result.content!r}")

    if payload.get("status") in {"DENIED", "ERROR"}:
        fail(
            f"decision was {payload['status']} "
            f"(reason: {payload.get('reason_code') or payload.get('message')})"
        )

    if not middleware.proofs:
        fail("no proof artifact emitted")

    proof = middleware.proofs[0]
    print()
    print("OK")
    print("  decision       = ALLOW")
    print(f"  proof_id       = {proof.proof_id}")
    print(f"  signature      = {proof.authority_signature[:32]}...")
    print(f"  latency_ms     = {proof.latency_ms:.1f}")
    print(f"  constraints    = {proof.constraints_evaluated}")


if __name__ == "__main__":
    main()

File: packages/js/scripts/smoke-test-authority.mjs
#!/usr/bin/env node
/**
 * End-to-end smoke test against a deployed Nuggets authority backend.
 *
 * Mirrors scripts/smoke_test_authority.py. Confirms the TypeScript SDK can mint
 * an OIDC token, sign the agent_proof, reach the live authority endpoint, get an
 * ALLOW, and verify the returned proof. Exits 0 on ALLOW + emitted proof.
 *
 * Build first, then run with the same env contract as the Python smoke:
 *
 *   cd packages/js && npm run build
 *   export NUGGETS_AUTHORITY_URL="https://accounts.nuggets.life"
 *   export NUGGETS_OIDC_ISSUER_URL="https://auth.nuggets.life"
 *   export NUGGETS_AGENT_ID="did:web:auth.nuggets.life:..."
 *   export NUGGETS_CONTROLLER_ID="..."
 *   export NUGGETS_DELEGATION_ID="10"
 *   export NUGGETS_AGENT_PRIVATE_KEY="/path/to/agent-jwks.json"   # or PEM
 *   export NUGGETS_TOOL="your_tool_name"   # an in-scope allowed_capability
 *   export NUGGETS_TARGET="optional-in-scope-target"
 *   export NUGGETS_AMOUNT_MINOR="500"     # optional; minor units (e.g. 500 = £5.00)
 *   export NUGGETS_CURRENCY="GBP"         # optional; ISO-4217, uppercase
 *   node scripts/smoke-test-authority.mjs
 *
 * The handler is a no-op — this script NEVER executes a real payment. Amount/
 * currency are synthetic inputs to exercise ALLOW / ESCALATE / DENY routing.
 */
import { readFileSync } from "node:fs";
import { ToolMessage } from "@langchain/core/messages";
import { MiddlewareConfig, NuggetsAuthorityMiddleware } from "../dist/index.js";

function env(name, required = true) {
  const value = process.env[name];
  if (required && (!value || value.length === 0)) {
    console.error(`Missing required env var: ${name}`);
    process.exit(2);
  }
  return value;
}

function loadKey(raw) {
  // Accept a PEM string, a path to a PEM/JWK/JWKS file, or inline JSON.
  if (raw.trimStart().startsWith("-----BEGIN") || raw.trimStart().startsWith("{")) {
    return raw.trimStart().startsWith("{") ? JSON.parse(raw) : raw;
  }
  const content = readFileSync(raw, "utf8");
  return content.trimStart().startsWith("{") ? JSON.parse(content) : content;
}

const tool = env("NUGGETS_TOOL");
const target = env("NUGGETS_TARGET", false);
const amountMinor = env("NUGGETS_AMOUNT_MINOR", false);
const currency = env("NUGGETS_CURRENCY", false);

// Only supply an action-context resolver when a synthetic amount/currency is
// requested. Never guess money fields — the resolver is the sole source.
const actionContextResolver = (amountMinor || currency)
  ? () => ({
      ...(amountMinor ? { amount_minor: Number(amountMinor) } : {}),
      ...(currency ? { currency } : {}),
      ...(target ? { target } : {})
    })
  : undefined;

const config = new MiddlewareConfig({
  apiUrl: env("NUGGETS_AUTHORITY_URL"),
  oidcIssuerUrl: env("NUGGETS_OIDC_ISSUER_URL"),
  agentId: env("NUGGETS_AGENT_ID"),
  controllerId: env("NUGGETS_CONTROLLER_ID"),
  delegationId: env("NUGGETS_DELEGATION_ID"),
  agentPrivateKey: loadKey(env("NUGGETS_AGENT_PRIVATE_KEY")),
  actionContextResolver
});

const middleware = new NuggetsAuthorityMiddleware(config);
const request = {
  tool_call: { name: tool, args: target ? { target } : {}, id: "smoke-call" }
};
// No-op handler: it never performs a payment or any side effect.
const handler = async () =>
  new ToolMessage({ content: JSON.stringify({ status: "ok" }), tool_call_id: "smoke-call" });

console.log("NOTE: smoke handler is a no-op — it does NOT execute any payment.");
console.log(
  `→ ${config.apiUrl}${config.authorityEndpoint}  tool=${tool}` +
    `${target ? ` target=${target}` : ""}` +
    `${amountMinor || currency ? ` amount_minor=${amountMinor ?? "?"} currency=${currency ?? "?"}` : ""}`
);

const result = await middleware.wrapToolCall(request, handler);
const content = result instanceof ToolMessage ? String(result.content) : JSON.stringify(result);

let parsed;
try {
  parsed = JSON.parse(content);
} catch {
  parsed = { status: "ok" };
}

if (parsed.status === "PENDING_APPROVAL") {
  console.log(`⏸ ESCALATE — approval ${parsed.approval_id ?? "?"} (${parsed.reason_code ?? ""}); no tool executed, no payment made`);
  process.exit(0);
}

if (parsed.status === "DENIED" || parsed.status === "ERROR") {
  console.error(`✗ ${parsed.status}: ${parsed.reason_code ?? ""} ${parsed.message ?? ""}`);
  process.exit(1);
}

const proof = middleware.proofs.at(-1);
if (!proof) {
  console.error("✗ ALLOW but no proof artifact was emitted");
  process.exit(1);
}
console.log(`✓ ALLOW — proof ${proof.proof_id} (${proof.latency_ms.toFixed(0)}ms), verified against the authority's published keys`);
process.exit(0);
